core/types/transaction_conditional.go

```diff
@@ -0,0 +1,256 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// KnownAccounts represents a set of KnownAccounts
+type KnownAccounts map[common.Address]KnownAccount
+
+// EncodeRLP will encode the known account into rlp bytes
+func (ka KnownAccounts) EncodeRLP(w io.Writer) error {
+	type accountKV struct {
+		Key   common.Address
+		Value KnownAccount
+	}
+	accounts := make([]accountKV, 0, len(ka))
+	for k, v := range ka {
+		accounts = append(accounts, accountKV{Key: k, Value: v})
+	}
+	return rlp.Encode(w, accounts)
+}
+
+// DecodeRLP will decode the known account into rlp bytes
+func (ka *KnownAccounts) DecodeRLP(s *rlp.Stream) error {
+	type accountKV struct {
+		Key   common.Address
+		Value KnownAccount
+	}
+	var accounts []accountKV
+	if err := s.Decode(&accounts); err != nil {
+		return err
+	}
+
+	kamap := *ka
+	for _, account := range accounts {
+		kamap[account.Key] = account.Value
+	}
+	return nil
+}
+
+// KnownAccount allows for a user to express their preference of a known
+// prestate at a particular account. Only one of the storage root or
+// storage slots is allowed to be set. If the storage root is set, then
+// the user prefers their transaction to only be included in a block if
+// the account's storage root matches. If the storage slots are set,
+// then the user prefers their transaction to only be included if the
+// particular storage slot values from state match.
+type KnownAccount struct {
+	StorageRoot  *common.Hash `rlp:"nil"`
+	StorageSlots map[common.Hash]common.Hash
+}
+
+// UnmarshalJSON will parse the JSON bytes into a KnownAccount struct.
+func (ka *KnownAccount) UnmarshalJSON(data []byte) error {
+	var hash common.Hash
+	if err := json.Unmarshal(data, &hash); err == nil {
+		ka.StorageRoot = &hash
+		ka.StorageSlots = make(map[common.Hash]common.Hash)
+		return nil
+	}
+
+	var mapping map[common.Hash]common.Hash
+	if err := json.Unmarshal(data, &mapping); err != nil {
+		return err
+	}
+	ka.StorageSlots = mapping
+	return nil
+}
+
+// MarshalJSON will serialize the KnownAccount into JSON bytes.
+func (ka *KnownAccount) MarshalJSON() ([]byte, error) {
+	if ka.StorageRoot != nil {
+		return json.Marshal(ka.StorageRoot)
+	}
+	return json.Marshal(ka.StorageSlots)
+}
+
+// EncodeRLP will serialize the KnownAccount into rlp bytes
+func (ka KnownAccount) EncodeRLP(w io.Writer) error {
```

Comment:
I think a pointer-receiver here is better, since it embeds a map field, to not incur any accidental copy cost (assuming it's not optimized by Go, not sure)